src/pm-conversations/__tests__/pmConversationsModel-test.js

```diff
@@ -0,0 +1,144 @@
+/* @flow strict-local */
+import Immutable from 'immutable';
+
+import { usersOfKey, keyOfExactUsers, reducer } from '../pmConversationsModel';
+import * as eg from '../../__tests__/lib/exampleData';
+
+describe('usersOfKey', () => {
+  for (const [desc, ids] of [
+    ['self-1:1', []],
+    ['other 1:1 PM', [123]],
+    ['group PM', [123, 345, 567]],
+  ]) {
+    test(desc, () => {
+      expect(usersOfKey(keyOfExactUsers(ids))).toEqual(ids);
+    });
+  }
+});
+
+describe('reducer', () => {
+  const initialState = reducer(undefined, ({ type: eg.randString() }: $FlowFixMe));
```

Comment:
> this way of getting the initial/previous state, i.e., running a dummy action through the reducer starting from an `undefined` state.

Yep! One nice thing about this is that it exercises the API that Redux already requires a reducer to support -- in fact it's mimicking what Redux itself does to initialize the state on creating the store.


> In [612ec04](https://github.com/zulip/zulip-mobile/commit/612ec046b793e82cce634d035aea80eca377adfd), I renamed some `initialState`s to `prevState`. At first, I might have suggested that we make the same rename here for consistency. But I see that the `initialState` here really _is_ kind of an initial state

Yep. I do mean this as the *initial* state, in the same sense we use that name in our reducers: a given test case may or may not choose it as the "previous" state for e.g. applying the reducer with some action, but in any case `initialState` is the one constant state that is the one the given reducer initializes its state to.

In that commit 612ec046b793e82cce634d035aea80eca377adfd, a few of the things called `initialState` happened to actually equal the reducer's initial state, but most of them didn't; they were all local to a particular test, and what they had in common is that they were the starting state for their respective tests. So `prevState` was a much better name, to avoid the collision with the reducer's own actual initial state.

Reply:
> It looks like we've often just spelled out the initial/previous state literally, although in my current revision of #4390 I take advantage of our new `eg.makeMessagesState` to set the initial/previous state in the tests for `messagesReducer`. That approach lets me minimize the diff when we change over to `Immutable.Map` for `state.messages`, while making it easier to avoid the `Immutable.Map({ 1: 'foo' })` (vs. `Immutable.Map([[1, 'foo']])`) mistake.

Yeah. One thing I like about this is that it's a step toward making the tests not care about the internals of how we implement a particular part of the state, treating it more as a black box. That's good for being able to change the data structures; I think it also makes the tests more readable and informative, by having them act at a level that's more like how we want to think about what the app is doing: send events and other actions in, and select meaningful data out.


> I think your approach here would work just as well for both of those things, so I'd be happy to give it a try over there if you'd like. (In that branch there are a few uses of `eg.makeMessagesState` in other test files, like unreadHuddlesReducer-test, that should probably stay, to get those advantages.)

I think in fact a good way to think about `eg.makeMessagesState` may be as a shorthand for starting from the initial state and iterating through the given messages to add them. It's good to have such a shorthand, because it helps keep the tests short and focused on the interesting parts.

It should be exactly equivalent to that alternate version; for pure black-box testing, it could even be implemented that way. But I think it's fine to have it implemented directly, as it is -- it's simple, probably more efficient, and it's one central piece of code to update when we change the representation so it's not a burden there.

Reply:
> I noticed that we lose just a bit of coverage by not calling `deepFreeze` on state objects we pass to the reducer; in theory, I suppose, the reducer could get away with improperly mutating the object that holds `map` and `sorted`.

Yeah. I'm not sure our use of `deepFreeze` has ever actually caught a bug -- I think that kind of bug is largely pretty avoidable just by writing the code in the style we're accustomed to writing it in. So the approach I've been inclined to is to use it in central places like `exampleData.js`, where one call can cover a large swath of tests (see also 45308e7716b6e04f1d02cbf81456c05efc9559ef); and to generally skip it in individual test files, to avoid the clutter.